[Profiler] Handle Tensor Sizes/Strides Parsing Error

sraikund16 opened the pull request:
Summary:
Currently some jobs are encountering the following trace, P1539415198. This suggests that when we are parsing through tensors the path is prone to encountering an invalid address. This is is possibly occurring because for some reason the sizes() and strides() of a Tensor seem to not be of the same dimensions. We assume such when iterating through the shapes to get the Ivalue generator. When browsing some of the tensor implementations, I found that some of the size and stride paths are different which could be the cause of this issue. Regardless, the profiler should be flexible enough to handle such issues without bringing down the whole main thread.

If the crashes still persist, it will still give us a data point as to where they are occurring and we can rule out the strides/sizes as the culprit

Test Plan: This change doesn't break anything in the happy path, just makes sure the bad path is not exited abruptly. We should use this in order to debug what the events are having mismatching dimensions between sizes and strides.

Differential Revision: D62008788


cc @robieta @chaekit @aaronenyeshi @guotuofeng @guyang3532 @dzhulgakov @davidberard98 @briancoutinho @sanrise

aaronenyeshi commented on torch/csrc/profiler/collection.cpp:
```diff
@@ -33,11 +33,18 @@ RawTensorMetadataBase::RawTensorMetadataBase(const at::Tensor& t)
+      "Tensor has mismatching sizes and strides. Sizes: ",
+      t.sizes(),
+      " Strides: ",
+      t.strides());
```
did you mean to print the sizes and strides rather than their dimensions? Seems like we're checking dims here

aaronenyeshi commented on torch/csrc/profiler/collection.cpp:
```diff
@@ -181,14 +188,29 @@ auto InputOutputEncoder::getIValueGenerator(const IOType& io_type) {
+        for (C10_UNUSED const auto _ : c10::irange(raw_metadata.stride_dim_)) {
+          if (tensor_size_strides_it.exhausted()) {
+            LOG(WARNING)
+                << "Expected Tensor Strides mismatch with raw Tensor metadata. Reported shapes may be innacurate!";
```
typo: inaccurate

aaronenyeshi commented on torch/csrc/profiler/collection.cpp:
```diff
@@ -181,14 +188,29 @@ auto InputOutputEncoder::getIValueGenerator(const IOType& io_type) {
-      for (C10_UNUSED const auto _ : c10::irange(raw_metadata.dim_)) {
+      if (tensor_metadata_it.exhausted()) {
+        LOG(WARNING)
+            << "Tensor metadata exhaused prematurely. Reported shapes may be innacurate!";
```
typo: inaccurate, exhausted

aaronenyeshi commented on torch/csrc/profiler/collection.cpp:
```diff
@@ -181,14 +188,29 @@ auto InputOutputEncoder::getIValueGenerator(const IOType& io_type) {
+      for (C10_UNUSED const auto _ : c10::irange(raw_metadata.size_dim_)) {
+        if (tensor_size_strides_it.exhausted()) {
+          LOG(WARNING)
+              << "Expected Tensor Size mismatch with raw Tensor metadata. Reported shapes may be innacurate!";
```
typo: inaccurate

aaronenyeshi reviewed (approved):
Looks great! Added a few nits

sraikund16 committed:
[Profiler] Handle Tensor Sizes/Strides Parsing Error (#134862)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134862

Currently some jobs are encountering the following trace, P1539415198. This suggests that when we are parsing through tensors the path is prone to encountering an invalid address. This is is possibly occurring because for some reason the sizes() and strides() of a Tensor seem to not be of the same dimensions. We assume such when iterating through the shapes to get the Ivalue generator. When browsing some of the tensor implementations, I found that some of the size and stride paths are different which could be the cause of this issue. Regardless, the profiler should be flexible enough to handle such issues without bringing down the whole main thread.

Test Plan: This change doesn't break anything in the happy path, just makes sure the bad path is not exited abruptly. We should use this in order to debug what the events are having mismatching dimensions between sizes and strides.

Reviewed By: aaronenyeshi

Differential Revision: D62008788

kit1980 commented:
@pytorchbot merge -f "Merged internally, the errors are unrelated No matching distribution found for numpy==2.0.0rc1"
